### PR TITLE
Add Multiple Tower Support in Condenser Entering Temp Reset Setpoint Managers

### PR DESCRIPTION
**Pull request by Myoldmopar**
This PR will ultimately be the implementation of allowing multiple towers in the CondenserEnteringTemp reset setpoint managers.  As of right now, I have done some restructuring of the SetPointManager code to lean it toward OO patterns and some other minor cleanup.  
- [x] Verify initial changes have no-diffs
- [x] Provide NFP to team
- [x] Provide design document
- [x] Draft code push
- [x] Code review
- [x] Final code submission
- [x] Documentation changes

### CODE REVIEW

**Comment by Myoldmopar** [comment]
Hmmmm, so I'm seeing a potential issue in the ideal condenser reset manager.  I'll try to tell the story with lines of code here.

During init, we search out for the condenser loop that contains the setpoint node defined in the setpoint manager object.  No problem.  Once we've identified that loop, we then loop over the demand side branches [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3683) with `BranchNum` and components [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3684) with `CompNum`.  We check the `TypeOf_Num` for that component and if it is a chiller, we start doing additional mining to pick up the chiller and pump branch, comp.  But within the same loop, we then set the following variables:

```
IdealCondEntSetPtMgr( SetPtMgrNum ).TypeNum = TypeNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum = LoopNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum;
IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum = BranchNum;
```

At this point, `CompNum` is still assigned to the one checked previously that was used to identify the chiller component.  Yet it is being assigned to the `TowerNum`.  This is probably passing under the radar because there is usually one component per branch.  But the bigger point is that `BranchNum` has the same effect, and is being stored in `CondBranchNum`.  [Later](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L6657) in the setpoint manager operation, this variable is then assigned to a local variable named `TowerBranchNum`, and used to look up the meters.  This appears to be blatantly wrong.  As I see it, the tower energy is not being used in searching for an optimal energy use, and the chiller branch num is being double counted.

Can someone spare a minute to verify I'm not missing something obvious?  Because if I'm right, this means this setpoint manager has been wrong from the [beginning](https://github.com/NREL/EnergyPlus/blame/develop/src/EnergyPlus/SetPointManager.cc#L3722).

FYI: @EnergyArchmage @mjwitte @rraustad @lgu1234

**Comment by rraustad** [comment]
It looks to me as if the 3 variables together should point right to the 
chiller.

PlantLoop( LoopNum ).LoopSide( \* ).Branch( BranchNum ).Comp( CompNum )

So I am not quite sure what this line is supposed to do since it points 
to a chiller:
TowerBranchNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum;

Is there another loop like this somewhere looking for a tower? If there 
is, it would point to the tower for that SetPtMgrNum

On 8/20/2015 1:34 PM, Edwin Lee wrote:

> Hmmmm, so I'm seeing a potential issue in the ideal condenser reset 
> manager. I'll try to tell the story with lines of code here.
> 
> During init, we search out for the condenser loop that contains the 
> setpoint node defined in the setpoint manager object. No problem. Once 
> we've identified that loop, we then loop over the demand side branches 
> here 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3683 
> with |BranchNum| and components here 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L3684 
> with |CompNum|. We check the |TypeOf_Num| for that component and if it 
> is a chiller, we start doing additional mining to pick up the chiller 
> and pump branch, comp. But within the same loop, we then set the 
> following variables:
> 
> |IdealCondEntSetPtMgr( SetPtMgrNum ).TypeNum = TypeNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum = LoopNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum; 
> IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum = BranchNum; |
> 
> At this point, |CompNum| is still assigned to the one checked 
> previously that was used to identify the chiller component. Yet it is 
> being assigned to the |TowerNum|. This is probably passing under the 
> radar because there is usually one component per branch. But the 
> bigger point is that |BranchNum| has the same effect, and is being 
> stored in |CondBranchNum|. Later 
> https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/SetPointManager.cc#L6657 
> in the setpoint manager operation, this variable is then assigned to a 
> local variable named |TowerBranchNum|, and used to look up the meters. 
> This appears to be blatantly wrong. As I see it, the tower energy is 
> not being used in searching for an optimal energy use, and the chiller 
> branch num is being double counted.
> 
> Can someone spare a minute to verify I'm not missing something 
> obvious? Because if I'm right, this means this setpoint manager has 
> been wrong from the beginning 
> https://github.com/NREL/EnergyPlus/blame/develop/src/EnergyPlus/SetPointManager.cc#L3722.
> 
> FYI: @EnergyArchmage https://github.com/EnergyArchmage @mjwitte 
> https://github.com/mjwitte @rraustad https://github.com/rraustad 
> @lgu1234 https://github.com/lgu1234
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/NREL/EnergyPlus/pull/5088#issuecomment-133088107.

## 

Richard Raustad
Florida Solar Energy Center
1679 Clearlake Road
Cocoa, FL 32922
Ph: (321)638-1454
http://www.fsec.ucf.edu/en/

Program Director
Electric Vehicle Transportation Center
http://evtc.fsec.ucf.edu/

**Comment by Myoldmopar** [comment]
Yeah, that's what I'm seeing too @rraustad.  That line I showed above is the only assignment of the `CondBranchNum` variable, so I think that is a blatant bug.  Crap, now I'm torn as to whether to include the bug fix here or do it separate.  Alright, well I'll fix it either way.

**Comment by rraustad** [comment]
This should be IdealCondEntSetPtMgr().CompNum (or since it's a chiller, ChillerNum):

IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum = CompNum;

The chiller index is stored here:
ChillerNum = IdealCondEntSetPtMgr( SetPtMgrNum ).ChillerIndexPlantSide;

And since the for loops only test for chillers, why are these even there:

TowerLoopNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondLoopNum;
TowerBranchNum = IdealCondEntSetPtMgr( SetPtMgrNum ).CondBranchNum;
TowerNum = IdealCondEntSetPtMgr( SetPtMgrNum ).TowerNum;

Oh, I see, the tower loop is just above the chiller loop. The only problem is they never stored the data (no assignments).

**Comment by rraustad** [comment]
You should probably have:
IdealCondEntSetPtMgr().ChillerCompNum 
IdealCondEntSetPtMgr().TowerCompNum 

since there can be a SPM for each of these, or one for one and not another. That way you'd know what component the SPM controlled.

**Comment by Myoldmopar** [comment]
@rraustad 

> since there can be a SPM for each of these

I'm not sure I get what you mean here.  This is a pretty high level setpoint manager that sets the condenser loop setpoint by querying the current state of the chiller, tower, and both pumps.  

Ultimately we will need some representation of the following:
- Scalars: ChillerBranchNum, ChillerCompNum
- Vectors: TowerBranchNums, TowerCompNums (since now allowing multiple towers, better ways to store multiple of these than separate arrays...)

File: doc/tools/create_changelog.py [hunk 1]
```diff
@@ -68,6 +68,16 @@ def usage():
...
+
+    # we need to skip very low numbers of pull requests, for example:
+    # - a user wants to contribute a change to E+, so they create a fork/branch
+    # - their operations result in a pull request into their own repo, so the counting starts at #1...
+    # we're at like 5000+, so if we just skip anything less than 1000, we'll be good.
+    # And look, I am even using lambdas to prove I don't hate them
+    expected_good_num = lambda n : int(pr_num) < 1000
+    if expected_good_num(pr_num):
+        continue
+
```
Comment (Myoldmopar) [inline doc/tools/create_changelog.py:80]: The above is an unrelated fix to clean up the changelog error handling; it doesn't actually affect the output of the changelog at this time.

File: src/EnergyPlus/SetPointManager.cc [hunk 2]
```diff
@@ -6686,16 +6639,19 @@ namespace SetPointManager {
...
 				// Get the chilled water pump energy consumption
-				ChilledPumpEnergy = GetInternalVariableValue( IdealCondEntSetPtMgr( SetPtMgrNum ).ChlPumpVarType, IdealCondEntSetPtMgr( SetPtMgrNum ).ChlPumpVarIndex );
+				ChilledPumpEnergy = GetInternalVariableValue( this->ChlPumpVarType, this->ChlPumpVarIndex );
 
 				// Get the cooling tower fan energy consumption
-				TowerFanEnergy = GetInternalVariableValue( IdealCondEntSetPtMgr( SetPtMgrNum ).ClTowerVarType, IdealCondEntSetPtMgr( SetPtMgrNum ).ClTowerVarIndex );
+				TowerFanEnergy = 0;
+				for ( int i = 1; i <= this->numTowers; i++ ) {
+					TowerFanEnergy += GetInternalVariableValue( this->ClTowerVarType( i ), this->ClTowerVarIndex( i ) );
+				}
```
Comment (Myoldmopar) [inline src/EnergyPlus/SetPointManager.cc]: Now loops over all towers on the loop to get the total energy use.  This loop is the core of this feature.

File: src/EnergyPlus/SetPointManager.cc [hunk 3]
```diff
@@ -7076,45 +6992,30 @@ namespace SetPointManager {
...
+			UnitsStrings.allocate( NumVariables );
+			ResourceTypes.allocate( NumVariables );
+			EndUses.allocate( NumVariables );
+			Groups.allocate( NumVariables );
+			Names.allocate( NumVariables );
+
+			GetMeteredVariables( TypeOfComp, NameOfComp, VarIndexes, VarTypes, IndexTypes, UnitsStrings, ResourceTypes, EndUses, Groups, Names, NumFound );
+			this->ClTowerVarType.push_back( VarTypes( 1 ) );
+			this->ClTowerVarIndex.push_back( VarIndexes( 1 ) );
+
```
Comment (Myoldmopar) [inline src/EnergyPlus/SetPointManager.cc:7126]: This loop gets the var type and var index for multiple towers on the condenser loop.

**Comment by Myoldmopar** [comment]
Well, if you don't count the new idfs I added, I am currently at a net of -781 lines of code with this PR.  I need to do a little more testing, and update the design doc with my current plans to complete it so I can put that out to review.  I know I have already done some coding, but I felt I needed to do that to determine the design.  [:egg:, :chicken:].first?

**Comment by mjwitte** [comment]
@Myoldmopar So, why don't the results change for CoolingTower_VariableSpeed_IdealCondEntTempSetpoint_MultipleTowers?  Ran this file as an annual simulation before and after these changes and the ABPUS report is identical.

**Comment by mjwitte** [comment]
Found that CoolingTower_VariableSpeed_IdealCondEntTempSetpoint_MultipleTowers test file didn't really have two towers.  Once this is updated, this can go in.

**Comment by Myoldmopar** [comment]
@mjwitte I updated the idf for multiple towers, pulled in develop, rebuilt, and re-ran.  I got good results.  I updated the [final section](https://github.com/NREL/EnergyPlus/blob/94006714-MultipleTowers/design/FY2015/CondenserResetWithMultipleTowers.md#modifications-and-testing-of-the-ideal-condenser-setpoint-manager) of the design doc with improved discussion.  I think this is ready to go in if you don't see any other issues.

**Comment by Myoldmopar** [comment]
And for the EnergyPlus source itself, I still ended up with a net lines of code added = -990.  w00t.

**Comment by mjwitte** [comment]
That didn't change anything.  For the ideal reset manager, the two-tower case still uses 2x heat rejection electricity compared to the single-tower case.

**Comment by Myoldmopar** [comment]
OK, I'm about ready to wash my hands of this.  It's late in the game and I'm in the mood to fix bugs.  We can chat on the sizing call about whether to push forward or not.

**Comment by mjwitte** [comment]
Once CI comes back with a couple more, I'll merge this and post a followup issue #5207
